File: Calcolatrice_C++/Model/vertice.cpp
#include "vertice.h"

Vertice::Vertice(): Punto(), info(0) {}

Vertice::Vertice(double x, double y, double i) : Punto(x, y), info(i) {}

Vertice::Vertice(const Punto &p, double i): Punto(p), info(i) {}

double Vertice::getInfo()const
{
    return info;
}

void Vertice::setInfo(double i)
{
    info = i;
}

bool Vertice::operator==(const Punto& p)const
{
    if(typeid(p)==typeid(const Vertice&) && Punto::operator==(p))
    {
        const Vertice& v = static_cast<const Vertice&>(p);
        return info==v.info;
    }
    else
        return false;
}

bool Vertice::operator!=(const Punto& p)const
{
    return !operator==(p);
}

string Vertice::toString() const
{
    stringstream buffer;
    buffer << *this;
    return buffer.str();
}

double AngoloVertici(const Vertice &a, const Vertice &b, const Vertice &c)
{ // Calcola l'angolo del vertice B, compreso tra A e C
    Punto ab( b.getX() - a.getX(), b.getY() - a.getY() );
    Punto cb( b.getX() - c.getX(), b.getY() - c.getY() );

    double dot = (ab.getX() * cb.getX() + ab.getY() * cb.getY() );
    double cross = (ab.getX() * cb.getY() - ab.getY() * cb.getX());

    double alpha = atan2(cross, dot);
    return (alpha * 180 / M_PI);
}

ostream& operator << (ostream& os, const Vertice& v)
{
    Punto p = v;
    return os<< "Vertice: (" << p.getX() << "," << p.getY()
             << ") " << v.getInfo();
}

File: Calcolatrice_Java/src/vertice/Vertice.java
package vertice;

import vertice.Punto;

public class Vertice extends Punto{ 
	private double info;
	
	public Vertice(Punto p, double info){
		super(p);
		this.info=info;
	}
	public Vertice(Punto p){
		this(p, 0);
	}
	public Vertice(){
		super(0,0);
		this.info=0;
	}

	protected Vertice(Vertice v){
		super(v);
		this.info = v.getInfo();
	}
	
	public double getInfo() {
		return info;
	}
	public void setInfo(double info) {
		this.info = info;
	}
	
	public static double AngoloVertici(final Vertice a, final Vertice b, final Vertice c){
		// Calcola l'angolo del vertice B, compreso tra A e C
	    Punto ab = new Punto( b.getX() - a.getX(), b.getY() - a.getY() );
	    Punto cb = new Punto( b.getX() - c.getX(), b.getY() - c.getY() );

	    double dot = (ab.getX() * cb.getX() + ab.getY() * cb.getY() );
	    double cross = (ab.getX() * cb.getY() - ab.getY() * cb.getX());

	    double alpha = Math.atan2(cross, dot);
	    return (alpha * 180 / Math.PI);
	}
	
	@Override
	public String toString() {
		return "Vertice: (" + this.getX() + "," + this.getY() + ") " + info;
	}
	
	@Override
	public boolean equals(Object obj) {
		if(!(obj instanceof Vertice)) 
			return false;

		if(this == obj) 
			return true;

		Vertice other = (Vertice)obj;
		return super.equals(other) && other.info==info;
	}
	
	@Override
	public Vertice clone(){ 
		return new Vertice(this);
	}
	
}





